### PR TITLE
[RadioGroup] Add `disabled` prop to `Root`

### PR DESCRIPTION
**Pull request by YarnSphere**
### Description

Allows setting `disabled` on the root `RadioGroup`, setting all `Radio` items as `disabled`.

The `RovingFocusGroup` has be altered to become unfocusable when it has no focusable items.

### CODE REVIEW

**Comment by YarnSphere** [comment]
My reasoning was 2-fold:
- At least one other library I was testing (Chakra) does this, not that this means much.
- The current behaviour is simply a "shorthand" to manually setting `disabled` on each item, so it made sense to be able to override on individual items if needed.

I agree with your reasoning, though. I'll change the implementation to do that. I'll further set `data-disabled` on the root, then, and I was thinking of adding a `tabIndex={-1}` to the root as well, what are your thought on this?

**Comment by benoitgrelard** [comment]
From Discord:

> @benoit Testing the disabled on the radio-group's root, I noticed something on the current radio-groups that I'm not sure is wanted: why does the group itself have tabIndex="0"? This means that the group will catch focus even if all radios within are disabled.

> In a "vanilla" radio group this doesn't seem to happen. Is this behaviour correct?
just tested a bit more and this might make sense, it seems like in a "vanilla" group if the currently selected radio is disabled, it doesn't let me tab into the whole group because it tries to tab into the selected one only 

Yeah, that's how we handle focusing the correct item when tabbing through the page, focus lands on the group, and the groups throws it back to where it knows it should be.

That does mean that in this case, when the whole group is disabled here, we need to override that `tabIndex`.
I think something like this to only override when we need and otherwise leave it to what is set by `RovingFocusGroup`:

```diff
<RovingFocusGroup.Root
  asChild
  {...rovingFocusGroupScope}
  orientation={orientation}
  dir={direction}
  loop={loop}
+  {...(disabled ? { tabIndex: -1 } : {})}
>
```

**Comment by YarnSphere** [comment]
This still doesn't cover the case where all items are manually set to `disabled`, in which case the group is still focusable. Would you like me to do something about this?

We could use the context to track the number of items vs number of disabled items and set `tabIndex` to `-1` when all items are disabled.

File: packages/react/radio-group/src/RadioGroup.tsx [hunk 1]
```diff
@@ -31,19 +31,21 @@ const useRadioScope = createRadioScope();
...
   name?: string;
   required: boolean;
+  disabled: boolean;
   value?: string;
   onValueChange(value: string): void;
 };
 
 const [RadioGroupProvider, useRadioGroupContext] =
-  createRadioGroupContet<RadioGroupContextValue>(RADIO_GROUP_NAME);
+  createRadioGroupContext<RadioGroupContextValue>(RADIO_GROUP_NAME);
```
Comment (benoitgrelard) [inline packages/react/radio-group/src/RadioGroup.tsx:40]: 👍

File: packages/react/radio-group/src/RadioGroup.tsx [hunk 2]
```diff
@@ -125,6 +129,7 @@ const RadioGroupItem = React.forwardRef<RadioGroupItemElement, RadioGroupItemPro
   (props: ScopedProps<RadioGroupItemProps>, forwardedRef) => {
     const { __scopeRadioGroup, disabled, ...itemProps } = props;
     const context = useRadioGroupContext(ITEM_NAME, __scopeRadioGroup);
+    const isDisabled = disabled ?? context.disabled;
```
Comment (benoitgrelard) [inline packages/react/radio-group/src/RadioGroup.tsx:132]: Yeah so I think this should be the other way around.

**Comment by benoitgrelard** [comment]
> This still doesn't cover the case where all items are manually set to disabled, in which case the group is still focusable. Would you like me to do something about this?

That's a great point!
It looks like perhaps a better place to deal with this would be in `RovingFocusGroup` itself, if all items are "non-focusable" then we'd remove the `tabIndex` requirement from the top.

Then it would just work automatically for radio group here, because all the items would be made disabled/thus non-focusable from `RovingFocusGroup` POV.

**Comment by YarnSphere** [comment]
> It looks like perhaps a better place to deal with this would be in RovingFocusGroup itself, if all items are "non-focusable" then we'd remove the tabIndex requirement from the top.

I've pushed the other changes for now, I'll try to have a go at implementing that.

**Comment by YarnSphere** [comment]
@benoitgrelard, I had a go at it and I don't think it's an easy change.

I think we'd have to actively track the `disabled` state across items, which would likely involve changes in all components using roving focus.

I tried to use `getItems`, check if none are `focusable`, and set `tabIndex={-1}` in such case; however, this needs to run whenever there are changes to the disabled state of individual items, when new items are added to the group, etc. Either that or we'd want to run this _just before_ the user tabs into the group, which I don't think is possible.

Anyway, unless I'm missing something, I suggest we land this as-is, as it's still a strict improvement over the current behaviour (at least with this implementation setting `disabled` on the group does prevent focusing it).

**Comment by YarnSphere** [comment]
> It looks like perhaps a better place to deal with this would be in RovingFocusGroup itself, if all items are "non-focusable" then we'd remove the tabIndex requirement from the top.

> Then it would just work automatically for radio group here, because all the items would be made disabled/thus non-focusable from RovingFocusGroup POV.

The above is implemented.

This PR should be ready for review. The only limitation of the current implementation is that if all `focusable` items are also hidden (`display: none`), then the group will still be focusable. I think that handling this edge case reliably would be quite difficult and probably not worth it.

**Review by andy-hook — COMMENTED** [review]
Sorry for the delay, I had a couple of additional thoughts

> A limitation of the current implementation is that if all focusable items are also hidden (display: none), then the group will still be focusable. I think that handling this edge case might not be worth it.

  I think this is fine, I can’t imagine it being very common.

File: packages/react/roving-focus/src/RovingFocusGroup.tsx [hunk 3]
```diff
@@ -119,6 +124,8 @@ const RovingFocusGroupImpl = React.forwardRef<
   const handleEntryFocus = useCallbackRef(onEntryFocus);
   const getItems = useCollection(__scopeRovingFocusGroup);
   const isClickFocusRef = React.useRef(false);
+  const [focusableStates, setFocusableStates] = React.useState<FocusableStates>({});
+  const noneFocusable = Object.values(focusableStates).every((state) => !state);
```
[the remark this answers is not in the document]
Comment (YarnSphere) [reply packages/react/roving-focus/src/RovingFocusGroup.tsx:128]: If I understand what you're saying correctly, how do we then update the set? Create a new instance every time an item is added/removed? Doesn't React track state by reference, meaning that we can't just do `set.add()`, `set.delete()`?

I can change the implementation to pass `onFocusableItemAdd`/`onFocusableItemRemove` via context, but I'm not sure that a set is a better data structure than an object here due to having to create a new instance every time.

I'll give this a go and see how it goes, maybe I'm misunderstanding how you'd use a set here.

[the remark this answers is not in the document]
Comment (YarnSphere) [reply packages/react/roving-focus/src/RovingFocusGroup.tsx:128]: Rewriting this, unless I'm missing something, I don't think there's even a need for tracking ids, so no data structure is needed; a simple counter with the number of focusable items seems enough. Will push an update soon.

File: packages/react/radio-group/src/RadioGroup.stories.tsx [hunk 4]
```diff
@@ -248,6 +248,21 @@ export const Chromatic = () => {
         </RadioGroup.Item>
       </RadioGroup.Root>
 
+      <h1>Disabled root</h1>
```
[the remark this answers is not in the document]
Comment (YarnSphere) [reply packages/react/radio-group/src/RadioGroup.stories.tsx:264]: Done.

File: packages/react/roving-focus/src/RovingFocusGroup.tsx (hunk 3 above)
[the remark this answers is not in the document]
Comment (YarnSphere) [reply packages/react/roving-focus/src/RovingFocusGroup.tsx]: Done.

**Comment by YarnSphere** [comment]
Rebased against `main`. This should be ready for review again.

**Comment by YarnSphere** [comment]
I think the release should be marked as `minor` instead of `patch` as it introduces a new API and new behaviour. I'll change the yarn versions file to reflect that.

**Comment by YarnSphere** [comment]
I marked all changes as `minor`, let me know if I should mark some dependencies as `patch` instead.

File: .yarn/versions/7e2a8b9b.yml [hunk 5]
```diff
@@ -0,0 +1,12 @@
+releases:
+  "@radix-ui/react-context-menu": minor
+  "@radix-ui/react-dropdown-menu": minor
+  "@radix-ui/react-menu": minor
+  "@radix-ui/react-radio-group": minor
```
Comment (andy-hook) [inline .yarn/versions/7e2a8b9b.yml:5]: Let's mark `react-radio-group` as `minor` and the rest can be `patch`

File: packages/react/roving-focus/src/RovingFocusGroup.tsx [hunk 6]
```diff
@@ -140,9 +143,11 @@ const RovingFocusGroupImpl = React.forwardRef<
         [setCurrentTabStopId]
       )}
       onItemShiftTab={React.useCallback(() => setIsTabbingBackOut(true), [])}
+      onFocusableItemAdd={React.useCallback(() => setFocusableItemsCount((n) => n + 1), [])}
+      onFocusableItemRemove={React.useCallback(() => setFocusableItemsCount((n) => n - 1), [])}
```
Comment (andy-hook) [inline packages/react/roving-focus/src/RovingFocusGroup.tsx:147]:
```suggestion
      onFocusableItemAdd={React.useCallback(() => setFocusableItemsCount((prevCount) => prevCount + 1), [])}
      onFocusableItemRemove={React.useCallback(() => setFocusableItemsCount((prevCount) => prevCount - 1), [])}
```

Nit: We typically prefer to explicitly label previous values in setters

File: packages/react/roving-focus/src/RovingFocusGroup.tsx [hunk 7]
```diff
@@ -198,10 +203,20 @@ const RovingFocusGroupItem = React.forwardRef<RovingFocusItemElement, RovingFocu
...
+    );
     const isCurrentTabStop = context.currentTabStopId === id;
     const getItems = useCollection(__scopeRovingFocusGroup);
 
+    React.useEffect(() => {
+      if (focusable) {
+        onFocusableItemAdd();
+        return () => onFocusableItemRemove();
+      }
+    }, [focusable, onFocusableItemAdd, onFocusableItemRemove]);
```
Comment (andy-hook) [inline packages/react/roving-focus/src/RovingFocusGroup.tsx:218]:
```suggestion
    const context = useRovingFocusContext(ITEM_NAME, __scopeRovingFocusGroup);
    const isCurrentTabStop = context.currentTabStopId === id;
    const getItems = useCollection(__scopeRovingFocusGroup);
    
    const { onFocusableItemAdd, onFocusableItemRemove } = context
    
    React.useEffect(() => {
      if (focusable) {
        onFocusableItemAdd();
        return () => onFocusableItemRemove();
      }
    }, [focusable, onFocusableItemAdd, onFocusableItemRemove]);
```

Nit: Small style thing but we prefer to keep `context` whole and destructure methods next to their effects when we need to please the linter

**Review by andy-hook — APPROVED** [review]
Good shout on the counter over structure, even simpler still! Just a few nits but otherwise this looks great to me.

Thanks!

File: packages/react/roving-focus/src/RovingFocusGroup.tsx (hunk 7 above)
Comment (YarnSphere) [reply packages/react/roving-focus/src/RovingFocusGroup.tsx]: Done.

File: packages/react/roving-focus/src/RovingFocusGroup.tsx (hunk 6 above)
Comment (YarnSphere) [reply packages/react/roving-focus/src/RovingFocusGroup.tsx]: Done.

File: .yarn/versions/7e2a8b9b.yml (hunk 5 above)
Comment (YarnSphere) [reply .yarn/versions/7e2a8b9b.yml]: Done.

**Pull request merged**